File: src/main/java/top/zbeboy/isy/aop/logging/LoggingAspect.java
package top.zbeboy.isy.aop.logging;


import lombok.extern.slf4j.Slf4j;
import org.aspectj.lang.JoinPoint;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.AfterThrowing;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Pointcut;
import org.springframework.core.env.Environment;
import top.zbeboy.isy.config.Workbook;

import javax.inject.Inject;
import java.util.Arrays;

/**
 * Aspect for logging execution of service and repository Spring components.
 * 运用切面输出日志信息.
 *
 * @author zbeboy
 * @version 1.0
 * @since 1.0
 */
@Slf4j
@Aspect
public class LoggingAspect {

    @Inject
    private Environment env;

    /**
     * 日志切面
     */
    @Pointcut("within(top.zbeboy.isy.service..*) || within(top.zbeboy.isy.web..*)")
    public void loggingPointcut() {
    }

    /**
     * 日志打印报错信息
     *
     * @param joinPoint
     * @param e
     */
    @AfterThrowing(pointcut = "loggingPointcut()", throwing = "e")
    public void logAfterThrowing(JoinPoint joinPoint, Throwable e) {
        if (env.acceptsProfiles(Workbook.SPRING_PROFILE_DEVELOPMENT)) {
            log.error("Exception in {}.{}() with cause = {} and exception {}", joinPoint.getSignature().getDeclaringTypeName(),
                    joinPoint.getSignature().getName(), e.getCause(), e);
        } else {
            log.error("Exception in {}.{}() with cause = {}", joinPoint.getSignature().getDeclaringTypeName(),
                    joinPoint.getSignature().getName(), e.getCause());
        }
    }

    /**
     * 日志切面环绕
     *
     * @param joinPoint
     * @return
     * @throws Throwable
     */
    @Around("loggingPointcut()")
    public Object logAround(ProceedingJoinPoint joinPoint) throws Throwable {
        if (log.isDebugEnabled()) {
            log.debug("Enter: {}.{}() with argument[s] = {}", joinPoint.getSignature().getDeclaringTypeName(),
                    joinPoint.getSignature().getName(), Arrays.toString(joinPoint.getArgs()));
        }
        try {
            Object result = joinPoint.proceed();
            if (log.isDebugEnabled()) {
                log.debug("Exit: {}.{}() with result = {}", joinPoint.getSignature().getDeclaringTypeName(),
                        joinPoint.getSignature().getName(), result);
            }
            return result;
        } catch (IllegalArgumentException e) {
            log.error("Illegal argument: {} in {}.{}()", Arrays.toString(joinPoint.getArgs()),
                    joinPoint.getSignature().getDeclaringTypeName(), joinPoint.getSignature().getName());

            throw e;
        }
    }
}

File: src/main/java/top/zbeboy/isy/aop/logging/LoggingAspect.kt
package top.zbeboy.isy.aop.logging

import org.aspectj.lang.JoinPoint
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.AfterThrowing
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.annotation.Pointcut
import org.slf4j.LoggerFactory
import org.springframework.core.env.Environment
import top.zbeboy.isy.config.Workbook
import java.util.*
import javax.inject.Inject

/**
 * Aspect for logging execution of service and repository Spring components.
 * 运用切面输出日志信息.
 *
 * @author zbeboy
 * @version 1.1
 * @since 1.0
 */
@Aspect
class LoggingAspect {

    private val log = LoggerFactory.getLogger(LoggingAspect::class.java)

    @Inject
    private lateinit var env: Environment

    /**
     * 日志切面
     */
    @Pointcut("within(top.zbeboy.isy.service..*) || within(top.zbeboy.isy.web..*)")
    fun loggingPointcut() {
    }

    /**
     * 日志打印报错信息
     *
     * @param joinPoint
     * @param e
     */
    @AfterThrowing(pointcut = "loggingPointcut()", throwing = "e")
    fun logAfterThrowing(joinPoint: JoinPoint, e: Throwable) {
        if (env.acceptsProfiles(Workbook.SPRING_PROFILE_DEVELOPMENT)) {
            log.error("Exception in {}.{}() with cause = {} and exception {}", joinPoint.signature.declaringTypeName,
                    joinPoint.signature.name, e.cause, e)
        } else {
            log.error("Exception in {}.{}() with cause = {}", joinPoint.signature.declaringTypeName,
                    joinPoint.signature.name, e.cause)
        }
    }

    /**
     * 日志切面环绕
     *
     * @param joinPoint
     * @return
     * @throws Throwable
     */
    @Around("loggingPointcut()")
    @Throws(Throwable::class)
    fun logAround(joinPoint: ProceedingJoinPoint): Any {
        if (log.isDebugEnabled) {
            log.debug("Enter: {}.{}() with argument[s] = {}", joinPoint.signature.declaringTypeName,
                    joinPoint.signature.name, Arrays.toString(joinPoint.args))
        }
        try {
            val result = joinPoint.proceed()
            if (log.isDebugEnabled) {
                log.debug("Exit: {}.{}() with result = {}", joinPoint.signature.declaringTypeName,
                        joinPoint.signature.name, result)
            }
            return result
        } catch (e: IllegalArgumentException) {
            log.error("Illegal argument: {} in {}.{}()", Arrays.toString(joinPoint.args),
                    joinPoint.signature.declaringTypeName, joinPoint.signature.name)

            throw e
        }

    }
}